File: python/hamming/hamming.py
"""
Exercism.io assignment - hamming.py
written by (GitHub):     cmdellinger

Usage:
   hamming.py <sequence1> <sequence2>

Compares the nucleotides in sequence1 and sequence2 and reports differences
also known as the Hamming distance.
"""

## ------
## Exercism.io function
## ------

def distance(string1, string2): #->
    """returns the differences (Hamming distance) between string1 and string2"""
    differences = 0

    for i in range(len(string1)):
        if string1[i] != string2[i]:
            differences += 1

    return differences

## ------
## Command-line implementation
## ------
if __name__ == '__main__':
    from docopt import docopt

    sequence1 = docopt(__doc__)['<sequence1>']
    sequence2 = docopt(__doc__)['<sequence2>']

    hamming_distance = distance(sequence1, sequence2)
    print("Hamming distance: %d" % hamming_distance)

File: ruby/hamming/hamming.rb
#### hamming ####
#
#   Hammming class with compute() class method.
#     compute() method determines hamming distance or difference between
#     two nucleotide sequences passed as strings.
#
#     Usage:  Hamming.compute(<"sequence1">, <"sequence2">)
#     Output: int(differences)

class Hamming
    def self.compute(string1 = "", string2 = "")
        # check that strings are the same length
        unless string1.length == string2.length
            raise ArgumentError.new("Sequences must be equal in length.")
        end
        # breaks strings into arrays of chars, zips them together,
        #   and counts different elements.
        return string1.chars.zip(string2.chars).count {|s1, s2| s1 != s2}
    end
end

#### BookKeeping Version ####
#
#   the version of the tests that are being run
#

module BookKeeping
    VERSION = 3 # Where the version number matches the one in the test.
end
